1_prep/src/munge_meteo.R

```diff
@@ -0,0 +1,66 @@
+#' @param outfile_template string representing the filepath at which to save
+#' the feather file output of the split netCDF data. The first `%s` is
+#' used as the placeholder for the `gcm_name`, the second is for the `gcm_date`,
+#' and the third is for the `cell_no`.
+#' @return names of cell/gcm/time-period specific feather files
+munge_nc_files <- function(gcm_nc, gcm_name, cell_no, gcm_date, outfile_template) {
+  # open netCDF file (gcm-specific) and pull data for that cell, for that time period
+  # TODO - actually pull time data for only one time period - pulling all for now
+  nc <- nc_open(gcm_nc)
+  time_dim <- ncvar_get(nc, "time")
+  cell_dim <- ncvar_get(nc, "gridcells")
+  data <- purrr::map_df(nc$var, function(var) {
+    ncvar_get(nc, var, start = c(1,cell_no), count = c(-1,1))
+  }) %>% mutate(time = nc.get.time.series(f = nc,time.dim.name = "time"),.before=1)
+
+  
+  # write to feather file, specific to gcm, cell, and time period
+  outfile <- sprintf(outfile_template, gcm_name, gcm_date, cell_no)
+  if(!dir.exists(dirname(outfile))) dir.create(dirname(outfile), recursive=TRUE)
+  arrow::write_feather(data, outfile)
+  
+  return(outfile)
+}
+
+#' @title build a lake-gcm-time-period crosswalk table
+#' @description Build a dplyr table with tar_grouping that has one row per
+#' Lake-GCM-time-period combination and columns for the (1) tar branch name 
+#' and (2) data hash for the meteo data for each combo. We can use this to run models.
+#' For the string extraction to work, cell_no must be at the END of the meteo filename
+#' @param meteo_feathers - names of the cell/gcm/time-period specific feather files
+#' @param lake_cell_xwalk - mapping of which lakes fall into which gcm cells
+#' @param gcm_names - names of the 6 GCMs
+#' @param gcm_dates - the three GCM time periods, defined by their bracketing years
+#' @return a dplyr tibble with one row per model run
+build_meteo_xwalk <- function(meteo_feathers, lake_cell_xwalk, gcm_names, gcm_dates){
+  # collapse the gcm_names, gcm_date, and cell_no vectors for use in string matching
+  gcm_name_list <- paste(gcm_names,collapse="|")
+  gcm_date_list <- paste(gcm_dates,collapse="|")
+  cell_no_list <- paste(unique(lake_cell_xwalk$cell_no),collapse= "|")
+  # Build tibble of meteo files, branches, hashes, gcm name, cell_no, and time_period
+  meteo_branches <- tibble(
+    meteo_fl = meteo_feathers,
+    meteo_branch = names(meteo_feathers),
+    meteo_data = map_chr(meteo_branch, ~ tar_meta(starts_with(.x))$data),
+    gcm = stringr::str_extract(meteo_fl, gcm_name_list),
+    time_period = stringr::str_extract(meteo_fl, gcm_date_list),
+    cell_no = as.numeric(str_extract(tools::file_path_sans_ext(meteo_fl), paste0(cell_no_list, "$")))
+  )
+  meteo_xwalk <- purrr::map_df(gcm_names, function(gcm_name) {tibble(gcm = gcm_name, time_period=gcm_dates)}) %>%
+    tidyr::expand(
+      nesting(gcm, time_period),
+      lake_cell_xwalk %>% select(site_id, ID, cell_no)) %>%
```

Comment:
I am completely inspired by your use of `tidyr::expand()` here. I usually use `base::expand.grid()`, which doesn't always return the data in a nice tidy format. Building off of this new found function, I think we may be able to simplify this first part of the `meteo_xwalk` by expanding the combinations all at once. The key is that we don't want to expand by anything in `lake_cell_xwalk %>% select(site_id, ID, cell_no)`, those relationships are set and should remain intact (as you already do). I _believe_ that is where that `nesting()` function comes in. I set up a small example to test this idea:

```r
gcm_dates <- c('1980_1999', '2040_2059', '2080_2099')
gcm_names <- c('ACCESS', 'GFDL', 'CNRM', 'IPSL', 'MRI', 'MIROC')

lake_cell_xwalk_test <- tibble(
  site_id = c("site1", "site2"), 
  ID = c("Navico_1", "Navico_2"),
  cell_no = c("24", "54")
)

all_model_combos_per_lake <- tidyr::expand_grid(
  nesting(lake_cell_xwalk_test),
  gcm_names, 
  gcm_dates
) 

nrow(all_model_combos_per_lake) # You get 36, which is 18 per site_id + ID + cell_no combo
```

If this checks out with you, I think this first part (before `dplyr::relocate()`) could be simplified to:
```r
meteo_xwalk <- tidyr::expand_grid(
  # Expand each lake (incl the site_id, ID, and cell_no with it)
  # for each combination of GCM + time period.
  nesting(select(lake_cell_xwalk, site_id, ID, cell_no)),
  gcm = gcm_names, 
  time_period = gcm_dates
) 
```

That was fun and now I have a few new functions in my toolbox - thanks 🙌

Reply:
Aha - thanks! I thought there must be a way to get rid of that first line, but wasn't landing on the right combo of `nesting()` , the lake_cell_walk, and the gcm and time_period vectors. Using `tidyr::expand_grid` is a nice solution. This simplified code works well!